File: src/main/java/com/leet/medium/MinimumDeletionsToMakeCharacterFrequenciesUnique.java
package com.leet.medium;

/**
 * 1647. Minimum Deletions to Make Character Frequencies Unique
 * https://leetcode.com/problems/minimum-deletions-to-make-character-frequencies-unique/
 * Medium
 * A string s is called good if there are no two different characters in s that have the same frequency.
 * Given a string s, return the minimum number of characters you need to delete to make s good.
 * The frequency of a character in a string is the number of times it appears in the string. For example, in the string "aab", the frequency of 'a' is 2, while the frequency of 'b' is 1.
 * Example 1:
 * Input: s = "aab"
 * Output: 0
 * Explanation: s is already good.
 * Example 2:
 * Input: s = "aaabbbcc"
 * Output: 2
 * Explanation: You can delete two 'b's resulting in the good string "aaabcc".
 * Another way it to delete one 'b' and one 'c' resulting in the good string "aaabbc".
 * Example 3:
 * Input: s = "ceabaacb"
 * Output: 2
 * Explanation: You can delete both 'c's resulting in the good string "eabaab".
 * Note that we only care about characters that are still in the string at the end (i.e. frequency of 0 is ignored).
 * Constraints:
 *     1 <= s.length <= 105
 *     s contains only lowercase English letters.
 */

import java.util.HashMap;
import java.util.Map;

public class MinimumDeletionsToMakeCharacterFrequenciesUnique {
  public int minDeletions(String s) {
    int[] chars = new int[26];
    for (int i = 0; i < s.length(); i++) {
      char ch = s.charAt(i);
      chars[ch - 'a']++;
    }

    Map<Integer, Integer> count = new HashMap<>();
    int max = 0;

    for (int i = 0; i < 26; i++) {
      max = Math.max(max, chars[i]);
      count.merge(chars[i], 1, Integer::sum);
    }

    int ans = 0;
    for (int i = max; i > 0; i--) {
      int ccount = count.getOrDefault(i, 0);
      if (ccount > 1) {
        ans += ccount - 1;
        count.merge(i - 1, ccount - 1, Integer::sum);
      }
    }

    return ans;
  }
}

File: src/main/java/com/leet/medium/MinimumDeletionsToMakeCharacterFrequenciesUnique.go
package main

// 100% faster
func minDeletions2(s string) int {
  count := make([]int, 26)
  for _, ch := range s {
    count[ch-'a']++
  }
  sort.Ints(count)
  ans := 0
  for i := len(count) - 2; i >= 0; i-- {
    if count[i+1] == 0 {
      if count[i] == 0 {
        break
      }
      ans += count[i]
      count[i] = 0
    } else if count[i] >= count[i+1] {
      ans += count[i] - count[i+1] + 1
      count[i] = count[i+1] - 1
    }
  }
  return ans
}

func minDeletions(s string) int {
  count := make([]int, 26)
  for _, ch := range s {
    count[ch-'a']++
  }
  sizeMap := map[int]int{}
  sizesMap := map[int]bool{}
  for _, size := range count {
    sizesMap[size] = true
    sizeMap[size]++
  }

  sizes := []int{}
  for size := range sizesMap {
    sizes = append(sizes, size)
  }

  sort.Ints(sizes)


  ans := 0
  for i := max; i > 0; i-- {
    sizeCount := sizeMap[i]
    if sizeCount > 1 {
      ans += sizeCount - 1
      sizeMap[i-1] += sizeCount-1
    }
  }

  return ans
}
